### PR TITLE
Dealias TypeBounds in uncurry

### LINKED ISSUE #11817: lambda + PartialFunction + specialization = "VerifyError: Bad local variable type"
this came up over at #7288, the original report there got fixed but this version remains broken even in 2.13.1

```
object SI_7288 {
  type Id[A] = A
  def f[A](pf: PartialFunction[A, Unit]): Unit = ()
  f[Id[Int]] { case _ => }
  def main(args: Array[String]): Unit = ()
}

Exception in thread "main" java.lang.VerifyError: Bad local variable type
Exception Details:
  Location:
    Main$$anonfun$1.applyOrElse(ILscala/Function1;)Ljava/lang/Object; @0: aload_1
  Reason:
    Type integer (current frame, locals[1]) is not assignable to reference type
```

### LINKED ISSUE #11884: Defdef vparams and symbol have inconsistent types
Me and @sjrd stumbled upon this bug while trying to fix https://github.com/scala-js/scala-js/issues/3953. It seems that `Defdef.vparams` and the respective `Defdef`'s symbol have inconsistent types.

### Steps for reproduction:
Clone the master branch https://github.com/scala-js/scala-js/commit/cc04be5b44d2a4dcfa0fcb663871d436ce8f918f of [Scala.js](https://github.com/scala-js/scala-js) locally and replace [`examples/helloworld/src/main/scala/helloworld/HelloWorld.scala`](https://github.com/scala-js/scala-js/blob/master/examples/helloworld/src/main/scala/helloworld/HelloWorld.scala) contents with the following:

```scala
import scala.language.higherKinds

sealed class StreamT[M[_]](val step: M[Step[StreamT[M]]])

sealed abstract class Step[S]

object Main {
  implicit class AnyOps[A](actual: A) {
    def mustMatch(f: PartialFunction[A, Boolean]): Unit = f.applyOrElse(???, ???)
  }

  type Id[A] = A

  def main(args: Array[String]): Unit = {
    (??? : StreamT[Id]).step mustMatch {
      case _ => true
    }
  }
}
```

Then, open an SBT shell and execute:
```
helloworld2_12/fastOptJS
```
Please note that you can replace

### PR DESCRIPTION
**Pull request by joroKr21**
fixes scala/bug#11817
fixes scala/bug#11884
alternative to #8720

### CODE REVIEW

**Comment by smarter** [comment]
Dealiasing can change the erasure of an intersection type and therefore its signature, so this change needs to be inspected very closely to guarantee binary compatibility. For example:
```scala
class foo extends scala.annotation.StaticAnnotation

class A
class B

class Test {
  // parameter type is erased to B
  def b_1(b: A with (B with A) @foo): Unit = {}
  type X = (B with A) @foo
  // parameter type is erased to A
  def a_2(a: A with X): Unit = {}
}

```

(I've been working on trying to emulate the Scala 2 intersection type erasure semantics in Dotty and it's pure madness because `intersectionDominator` takes non-erased types as input and then call things like `isNonBottomSubClass` which don't really have any well defined semantics on non-class types, here's the testcases I have so far: https://github.com/smarter/dotty/blob/cf528725bb21a4e63c74d2bd93787a13dc615b71/sbt-dotty/sbt-test/scala2-compat/types/scala2Lib/Api.scala and the corresponding implementation which I haven't polished/commented yet: https://github.com/smarter/dotty/blob/cf528725bb21a4e63c74d2bd93787a13dc615b71/compiler/src/dotty/tools/dotc/core/TypeErasure.scala#L377-L569)

**Comment by joroKr21** [comment]
It seems that this is a more invasive change - it didn't work. I agree it's a very sorry and unprincipled state of affairs. The deeper I look the more I despair. Now it seems that uncurry will dealias (even normalize) random methods (e.g. why PolyType?) and this affects erasure because of the other bug...

**Comment by joroKr21** [comment]
@smarter how can I turn your cases into an actual test? Print the trees after erasure?

**Comment by smarter** [comment]
I have a main for them here: https://github.com/smarter/dotty/blob/cf528725bb21a4e63c74d2bd93787a13dc615b71/sbt-dotty/sbt-test/scala2-compat/types/dottyApp/Main.scala#L40-L48 (you can ignore the part above that actually tries to call these methods, it's to test the interop, and it doesn't seem to work in scala 2 anyway because `null.asInstanceOf` throws a NullPointerException for some reason?)

**Comment by joroKr21** [comment]
> [error] java.io.IOException: Server returned HTTP response code: 403 for URL: https://scala-ci.typesafe.com/artifactory/scala-pr-validation-snapshots;build.timestamp=1596714086674/org/scala-lang/scala-library/2.13.4-bin-38edc7a-SNAPSHOT/scala-library-2.13.4-bin-38edc7a-SNAPSHOT.pom

**Review by lrytz — COMMENTED** [review]
I sort of grok what's going on. Would you mind putting in words your explanation what dealiasing in bounds changes, why it fixes the issues? 

Also we should add a test for the case discussed in https://github.com/scala/scala/pull/8720#issuecomment-640512903.

File: src/reflect/scala/reflect/internal/transform/UnCurry.scala [hunk 1]
```diff
@@ -37,87 +36,80 @@ trait UnCurry {
...
    *    Vector(tab, tb, tab)
    *  I think that difference is not the product of sentience but of randomness.
-   *  Let us figure out why it is and then change this method.
+   *  The reason for this are nested RefinedTypes that are flattened by tp.normalize.
+   *
+   *  And normalizing AnyRef leads to another single test failure: run/t2503.scala.
    */
-  private def expandAlias(tp: Type): Type = if (!tp.isHigherKinded) tp.normalize else tp
+  private def expandAlias(tp: Type): Type =
+    if (!tp.isHigherKinded && tp.typeSymbolDirect != AnyRefClass) tp.normalize else tp
```
Comment (lrytz) [inline src/reflect/scala/reflect/internal/transform/UnCurry.scala:44]: Meh i had a longer comment here that got lost...

Let me try again. I'm a little worried about the AnyRef special case.

I digged a bit and what breaks is `rebindSuper` in mixin failing to find the super call target.

```
trait SB[A] extends Buffer[A] {
  abstract override def update(n: Int, newelem: A): Unit = synchronized {
    super.update(n, newelem)
  }
}
new ArrayBuffer[Int] with SB[Int]
```

rebindSuper does
  - `acc.matchingSymbol(bcs.head, site)` where acc=mixin accessor for update, bcs.head = ArrayBuffer, site = ArrayBuffer[Int] with SB[Int]
  - the underlying problem is that `acc.tpe.asSeenFrom(site, acc.owner)` is `(idx: Int, elem: A): Unit` instead of `(idx: Int, elem: Int): Unit`
  - since the whole thing runs `exitingSpecialize` I tried various phases, and it seems the change is entering/exiting the fields phase. we get the right signature before fields. i have no idea why...

Comment (lrytz) [reply src/reflect/scala/reflect/internal/transform/UnCurry.scala:44]: This is just a long way of saying that this is pretty tricky territory...

**Comment by dwijnand** [comment]
@joroKr21 this one too, if you want to see it in 2.13.4.

File: src/reflect/scala/reflect/internal/transform/UnCurry.scala (hunk 1 above)
Comment (joroKr21) [reply src/reflect/scala/reflect/internal/transform/UnCurry.scala:44]: I wish I could understand why `AnyRef` vs `Object` makes a difference.

Comment (joroKr21) [reply src/reflect/scala/reflect/internal/transform/UnCurry.scala:44]: The weirdest thing is that we get the correct signature for `insertAll` but nor for `update` ...

**Comment by joroKr21** [comment]
I'm sorry I couldn't solve the problem with `AnyRef` and I don't think this should be merged without figuring that out. I think the proper fix is in `erasure`. If we can't do this because of binary compatibility than it might be better to just wait until Scala 3.
